services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,339 @@
+        for (Descriptor nestedType : type.getNestedTypes()) {
+          processType(nestedType, fd);
+        }
+      }
+
+      private void processFileDescriptor(FileDescriptor fd) {
+        String fdName = fd.getName();
+        Preconditions.checkState(!fileDescriptorsByName.containsKey(fdName),
+            "File name already used: %s", fdName);
+        fileDescriptorsByName.put(fdName, fd);
+        for (ServiceDescriptor service : fd.getServices()) {
+          processService(service, fd);
+        }
+        for (Descriptor type : fd.getMessageTypes()) {
+          processType(type, fd);
+        }
+        for (FieldDescriptor extension : fd.getExtensions()) {
+          processExtension(extension, fd);
+        }
+      }
+
+      /**
+       * Processes the service file descriptors. This prepares a list of the service names and
+       * indexes the file descriptors by name, symbol, and extension.
+       */
+      private void initFileDescriptorMaps() {
+        serviceNames = new HashSet<String>();
+        fileDescriptorsByName = new HashMap<String, FileDescriptor>();
+        fileDescriptorsBySymbol = new HashMap<String, FileDescriptor>();
+        fileDescriptorsByExtensionAndNumber = new HashMap<String, Map<Integer, FileDescriptor>>();
+
+        List<ServerServiceDefinition> services = server.getServices();
+
+        Queue<FileDescriptor> fileDescriptorsToProcess = new LinkedList<FileDescriptor>();
+        Set<String> seenFiles = new HashSet<String>();
+        for (ServerServiceDefinition service : services) {
+          io.grpc.ServiceDescriptor serviceDescriptor = service.getServiceDescriptor();
+          if (serviceDescriptor.getMarshallerDescriptor() instanceof ProtoFileDescriptorSupplier) {
+            FileDescriptor fileDescriptor =
+                ((ProtoFileDescriptorSupplier) serviceDescriptor.getMarshallerDescriptor())
+                    .getFileDescriptor();
+            String serviceName = serviceDescriptor.getName();
+            Preconditions.checkState(!serviceNames.contains(serviceName),
+                "Service already defined: %s", serviceName);
+            serviceNames.add(serviceName);
+            if (!seenFiles.contains(fileDescriptor.getName())) {
+              seenFiles.add(fileDescriptor.getName());
+              fileDescriptorsToProcess.add(fileDescriptor);
+            }
+          }
+        }
+
+        while (!fileDescriptorsToProcess.isEmpty()) {
+          FileDescriptor currentFd = fileDescriptorsToProcess.remove();
+          processFileDescriptor(currentFd);
+          for (FileDescriptor dependencyFd : currentFd.getDependencies()) {
+            if (!seenFiles.contains(dependencyFd.getName())) {
+              seenFiles.add(dependencyFd.getName());
+              fileDescriptorsToProcess.add(dependencyFd);
+            }
+          }
+        }
+      }
+
+      @Override
+      public void onNext(ServerReflectionRequest request) {
+        initFileDescriptorMaps();
```

Comment:
@carl-mastrangelo Breaking it into immutable and mutable data is fine, but that would require some method to differentiate between internal and dynamic services. The getServices() method on Server could return the internal services, and Server.getDynamicServices() (a new method) could return the dynamic ones?

@ejona86 had earlier mentioned that optimization of the reflection service wasn't particularly important at this point. Is it worthwhile differentiating between immutable and dynamic services in order to avoid this inefficiency in building the lookup tables?